sys/hoon.hoon

```diff
@@ -4264,296 +4270,477 @@
   |=  {{tab/@ edg/@} tac/tank}  ^-  wall
   (~(win re tac) tab edg)
 ::
+::  This library includes `plume`, the actual pretty printing logic,
+::  and a handful of utilities for constructing plums.
+::
+::  Generally, you'll just use `plume` like this:
+::
+::    ~(plume tall plum)  ::  Pretty print `plum` in tall mode.
+::    ~(plume falt plum)  ::  Pretty print `plum` in wide mode.
+::
+::  There is probably no reason to look at the utility routines unless
+::  you are writing something to generate `plum`s.
+::
+:: |^  :*  plume=plume
+::         fixed=fixed
+::         tall-fixed=tall-fixed
+::         tall-running=tall-running
+::         rune=rune
+::         simple-wide=simple-wide
+::         subtree=subtree
+::         sexp=sexp
+::      ==
+::
+::  This is the pretty-printer.  Use the `flat` arm to render a plum
+::  into a single line and use the `tall` arm to get a nice multi-line
+::  rendering that switches to wide mode if there's enough space.
+::
+::  For details about how this works and what exactly it does in various
+::  cases, take a look at the docs for `plum`, `plumfmt`, and at the
+::  docs on the arms of this door.
+::
 ++  plume
   |_  =plum
   ::
+  ::  An line, indented by `indent` spaces.
+  ::
+  +$  line  [indent=@ud text=tape]
+  ::
+  ::  An sequence of indented lines.
+  ::
+  +$  block  (list line)
+  ::
   ::  +flat: print as a single line
   ::
   ++  flat
+    ^-  wain
     text:linear
   ::
   ::  +tall: print as multiple lines
   ::
   ++  tall
     ^-  wain
     %+  turn  window
-    |=  [indent=@ud text=tape]
+    |=  line
     (crip (runt [indent ' '] text))
   ::
   ::  +adjust: adjust lines to right
   ::
   ++  adjust
-    |=  [tab=@ud =(list [length=@ud text=tape])]
-    (turn list |=([@ud tape] [(add tab +<-) +<+]))
-  ::  
+    |=  [tab=@ud =block]  ^-  ^block
+    (turn block |=([@ud tape] [(add tab +<-) +<+]))
+  ::
+  ::  Prepend `n` spaces to a tape.
+  ::
+  ++  prepend-spaces
+    |=  [n=@ t=tape]  ^-  tape
+    (runt [n ' '] t)
+  ::
+  ::  Append an element to the end of a list.
```

Comment:
I think this function should be moved into the generic list section. I've definitely wished we had it a few times.

Reply:
We might want to make sure it doesn't have a TMI issue while we're at it

Reply:
TMI issue?

Reply:
To be pedantic, while that points at the right intuition, `?~` is not the issue: wet gates are checked only for correctness, not completeness like dry gates, so `(|*(a=@ ?-(a %b a, %c a)) %b)` will give you `%b` just fine. If you try to use a non-nil sample for `^+` forever, on an output that *can* be `~`, you run into problems.